Replace log(1 + x) with log1p(x)

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92116
* #92115
* #92154
* __->__ #92114

`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

peterbell10 committed:
Replace log(1 + x) with log1p(x)

`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 0.35x speedup on CPU.

[ghstack-poisoned]

lezcano reviewed (commented):
OMG, I've been meaning to do this since forever! Thank you!

A couple other places:
Here it'd help to use it as the input is around 1 for small t

https://github.com/pytorch/pytorch/blob/b7cad020b59ec2c9b9d62a1e6ca78c5b53f186a1/torch/distributions/transforms.py#L827

https://github.com/pytorch/pytorch/blob/b7cad020b59ec2c9b9d62a1e6ca78c5b53f186a1/torch/distributions/exponential.py#L73

https://github.com/pytorch/pytorch/blob/b7cad020b59ec2c9b9d62a1e6ca78c5b53f186a1/torch/distributions/cauchy.py#L70

These two:
https://github.com/pytorch/pytorch/blob/b7cad020b59ec2c9b9d62a1e6ca78c5b53f186a1/torch/distributions/bernoulli.py#L116-L121

Every `log` hit in the file
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp

Two more hits at
aten/src/ATen/core/TransformationHelper.h
and one more at
aten/src/ATen/core/DistributionsHelper.h

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

[ghstack-poisoned]

peterbell10 commented:
@lezcano  I did see the distribution uses but not sure what our reproducibility guarantees are, so didn't want to touch it.

For the logit functions, I tried `log(x) - log1p(-x)` but it does degrade performance on CPU.

lezcano commented:
What about the last two files?

Also, cc @fritzo for confirmation on whether we can use `log1p` on distributions. I would say it'd be very much desirable to use it there as per the following discussion https://github.com/pytorch/pytorch/pull/91673#discussion_r1068790230.

We don't dispatch to the fast approximation of `log` in core, but still...

fritzo commented:
@lezcano thanks for doing this. Yes it seems fine to replace `log(1+x)` with `log1p(x)` in distributions.

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ngimel commented:
>We don't dispatch to the fast approximation of log in core, but still...

It's complicated. We don't for plain log function (it calls `::log` that gets resolved to normal log), but we do in a lot of distributions that call `at::log` (that calls fast approximation).

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -535,7 +535,7 @@ def binary_cross_entropy(
     #     "all elements of input should be between 0 and 1"
     # )
     loss = (target - 1) * torch.maximum(
-        torch.log(1 - self), self.new_full((), -100)
+        torch.log1p(-self), self.new_full((), -100)
     ) - target * torch.maximum(torch.log(self), self.new_full((), -100))
```
If it weren't for the `torch.maximum`s, this could be golfed as
```python
target * x.logit() + x.neg().log1p()
```

fritzo replied:
I'd assume the clamping is important, but I don't know why the author preferred `torch.maximum` over the cheaper `self.log().clamp(min=-100)`

ngimel replied:
Yeah replacing maximums with clamp_mins would avoid those `new_fulls`, but I think this is getting optimized away anyway.

lezcano reviewed (approved):
Just two notes, otherwise it LGTM

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on torch/distributions/transforms.py:
```diff
@@ -824,7 +824,7 @@ def _inverse(self, y):
         t = y_vec / (y_cumsum_vec).sqrt()
         # inverse of tanh
-        x = ((1 + t) / (1 - t)).log() / 2
+        x = (t.log1p() - t.neg().log1p()) / 2
```
I suspect the `t / (1 - t)` will lose us some precision, but I guess it's a good balance between performance and precision.

fritzo replied:
Correct, we've intentionally avoided log-of-ratio computations because they are numerically unstable.  We've preferred `log(x) - (-x).log1p()` or `.logit()` as numerically stable alternatives.

lezcano replied:
fair enough and good to know!

peterbell10 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1955,7 +1955,7 @@ Tensor binary_cross_entropy_target_backward(
     const c10::optional<Tensor>& weight,
     int64_t reduction) {
-  auto grad_target = (1. - self).log_().sub_(self.log());
+  auto grad_target = self.neg().log1p_().sub_(self.log());
```
Changed to `at::logit(self).neg()`

lezcano replied:
yup, even better!

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano reviewed (approved):
Fantastic PR! Feel free to merge whenever you see fit.
Edit. I didn't realise this was part of a stack, but yeah.

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Lint on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace log(1 + x) with log1p(x)"


`log1p` offers better precision near zero since `(1 + x) - 1` truncates any
values less than the float epsilon to zero. For `soft_margin_loss` this also
requires one fewer kernel invocation which for numel=1e7 gives me a 1.2x speedup
on CUDA and a 1.1x speedup on CPU.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
